can/interfaces/canalystii.py

```diff
@@ -69,132 +58,488 @@ class VCI_CAN_OBJ(Structure):
+        init_config.Timing1 = Timing1
+
+        self.init_config = init_config
+
+        status = CANalystII.VCI_InitCAN(
+            VCI_USBCAN2,
+            device,
+            channel,
+            byref(self.init_config)
+        )
+
+        if status != STATUS_OK:
+            self.shutdown()
+            raise CANalystIIStartupError(
+                "VCI_InitCAN Error, incorrect channel number?",
+                status
+            )
+
+        status = CANalystII.VCI_StartCAN(
+            VCI_USBCAN2,
+            device,
+            channel
+        )
+
+        if status != STATUS_OK:
+            logger.error(
+                "VCI_StartCAN Error ({0}), device:{1} "
+                "channel:{2}".format(status, device, channel)
+            )
+            logger.info('receiving failed to start, only able to transmit.')
 
-    def send(self, msg, timeout=None):
+    def board_info(self) -> Dict:
         """
+        Adapter Information
+
+        :return: dictionary containing the following keys
+
+            * `hardware_version`: `int`
+            * `firmware_version`: `int`
+            * `driver_version`: `int`
+            * `library_version`: `int`
+            * `irq`: `int`
+            * `num_can_interfaces`: `int`
+            * `serial_number`: `str`
+            * `hardware_type`: `str`
+
+        :rtype: dict
+        """
+        return {key: value for key, value in self._b_info.items()}
+
+    def send(self, msg: Message, timeout=None):
+        """
+        Send CAN Frame
 
         :param msg: message to send
+        :type msg: :class:`can.Message`
+
         :param timeout: timeout is not used here
-        :return:
+        :type timeout: None
+
+        :return: `True` if frame was sent else `False`
+        :rtype: bool
         """
-        extern_flag = 1 if msg.is_extended_id else 0
-        raw_message = VCI_CAN_OBJ(
-            msg.arbitration_id,
-            0,
-            0,
-            1,
-            msg.is_remote_frame,
-            extern_flag,
-            msg.dlc,
-            (c_ubyte * 8)(*msg.data),
-            (c_byte * 3)(*[0, 0, 0]),
+
+        can_objs = (VCI_CAN_OBJ * 1)()
+
+        can_objs[0].ID = msg.arbitration_id
+        can_objs[0].ExternFlag = int(msg.is_extended_id)
+        can_objs[0].SendType = 1
+        can_objs[0].RemoteFlag = int(msg.is_remote_frame)
+        can_objs[0].DataLen = msg.dlc
+        for j in range(msg.dlc):
+            can_objs[0].Data[j] = msg.data[j]
+
+        frames_sent = CANalystII.VCI_Transmit(
+            VCI_USBCAN2,
+            self.device,
+            self.channel,
+            byref(can_objs),
+            1
         )
 
-        if msg.channel is not None:
-            channel = msg.channel
-        elif len(self.channels) == 1:
-            channel = self.channels[0]
-        else:
-            raise ValueError("msg.channel must be set when using multiple channels.")
+        return bool(frames_sent)
```

Comment:
Raise an exception if frame could not be sent so the behavior is same as other interfaces.

Reply:
you are looking at a stale commit with the raising of a CanError

Reply:
The latest version of the send method seem to return a boolean. Do you mean that you have changed to raising an exception? In that case I can't see that change for some reason.

Reply:
I thought you were wanting me to change this
```python
else:
    raise ValueError("msg.channel must be set when using multiple channels.")
```
to a can error.

I can change the other to raising an exception if wanted. I would like to make a single point before I do, you may change your mind on it. Exception catching is expensive compared to boolean evaluation.

I used this script
```python
import timeit

code1 = """\
test = True

def send():
    return test

while not send():
    pass

while not send():
    pass

"""

code2 = """\
test = True

def send():
    if not test:
        raise RuntimeError()

try:
    send()
except RuntimeError:
    pass

"""

print(timeit.timeit(code1, number=1000000))
print(timeit.timeit(code2, number=1000000))
```

and these are the results for running each one 1000000 times
boolean = 0.337939405
exception = 0.42874530899999996

running this test 1000000 times may seem excessive, it's not. take a vehicle for instance, say there are 20 OBDII PIDS that I want to collect data from every 5 milliseconds. that's 4000 messages a second, 240000 messages a minute and after only 5 minutes of the application running we are at 1200000. with a time difference of 90 milliseconds that ends up costing 1.296 seconds every hour the application runs. after 24 hours it's at 31 seconds.

using the exception catching way you end up with a double whammy because you need to evaluate the returned value in order to decide whether or not you want to raise the exception nd in the user code exception catching needs to be put into place, where as we are not evaluating anything in the send method.

I know it seems trivial but it really adds up over time

Reply:
Perhaps so but that would be a major change to the API which many applications rely upon. That needs to be discussed in a separate issue.

Reply:
I understand, thought it worth mentioning.

When I first got into programming I was writing firmware for micro controllers, and because of the extremely limited resources available on those things code had to be written with resources being the #1 thing to pay attention to. Plus I hate it when an application does not have it's code optimized (cough.. firefox, cough..chrome) and is a resource hog when it shouldn't be. One of the reasons why I do not like Java, gobbles up memory and doesn't give it back to the OS when it is no longer in use.

I do what I can to keep things running as efficiently as possible and sometimes that requires writing it over and over again and lots of testing to get it dialed in for the proper balance of memory consumption vs CPU use.

Reply:
Same here, I also come from an embedded background, but sometimes I need to remind myself that it is a high level language after all and it is not designed to be fast.